Sources/Core/CompilerInfo.swift

```diff
@@ -0,0 +1,69 @@
+/// Describes the compiler; used as source of truth for conditional compilation.
+public struct CompilerInfo {
+
+  /// A version number with multiple parts; form: x(.y)*
+  public struct VersionNumber: Codable, Comparable {
+
+    public let parts: [Int]
+
+    public init(_ p: [Int]) {
+      parts = p
+    }
+
+    public static func < (lhs: Self, rhs: Self) -> Bool {
+      return lhs.parts.lexicographicallyPrecedes(rhs.parts)
+    }
+
+  }
+
+  /// The operating-system used when performing the compilation.
+  let os: String
+  /// The architecture of the machine doing the compilation.
+  let arch: String
+  /// The name of the compiler.
+  let compiler: String
+  /// The version of the compiler.
+  let compilerVersion: VersionNumber
+  /// The version of the Hylo language version we are targeting.
+  let hyloVersion: VersionNumber
+
+  /// We only need one instance of this struct, to represent the compiler information.
+  public static let instance = CompilerInfo()
+
+  private init() {
+    os = CompilerInfo.currentOS()
+    arch = CompilerInfo.currentArch()
+    compiler = "hc"
+    compilerVersion = VersionNumber([0, 1])
+    hyloVersion = VersionNumber([0, 1])
+  }
+
+  /// The current OS name.
```

Comment:
```suggestion
  /// The name of the operating system on which this function is run.
```
I feel like "current" is not as descriptive but don't feel strongly about it.